### DISCUSSION TITLE
Экспорт плейлиста - предложения по улучшению

### DISCUSSION THREAD

**Discussion by KisoidPoh095** [Chimildic/YaMuTools#16] [Идея] [answered]
Доброго времени суток!

Сегодня на странице экпорта (https://sites.google.com/view/yandex-music-to-text/) увидел сообщение, что Яндекс там что-то изменил и теперь надо ставить расширение...

Поставил расширение (Firefox 81.0.2 (64-битный)). Сделал экспорт. Вроде работает.

**Что порадовало:**
- теперь экспортируются приватные треки (те, которые я загрузил сам со своего компа)

**Что не порадовало:**
- сохраняется информация только о первичном исполнителе (у трека может быть несколько исполнителей, они отсортированы слева направо; под первичным я понимаю того, который левее всех). Т.к. редакция ЯМузыки достаточно часто косячит с выбором первичного исполнителя, считаю, что надо сохранять полный список имён. К тому же иногда название исполнителя, содержащее амперсанд (&) может волшебным образом разбиваться на двух исполнителей (например, "Dzihan & Kamien" превращается в "Dzihan","Kamien" и в результате экспорта будет только "Dzihan - название композиции")
- не сохраняется информация о редакции композиции (чей ремикс и т.д.). И это достаточно существенная потеря полезной информации, т.к. иногда ремиксы очень разнятся друг от друга и от оригинальной композиции
(я понимаю, что во время скробблинга эта информация тоже теряется, но хотелось бы хотя бы в экспортируемом плейлисте её сохранить...)

Таким образом, резюмируя вышенаписанное, старый вариант экспорта (который меня полностью устраивал, если не обращать внимание на отсутствие разделителей между именами исполнителей)
"Afterlife Max Essa - Makes Me Feel (Max Essa Remix)" превращается в "Afterlife - Makes Me Feel".

Очень хотелось бы вернуть как было =)

**UPD**

Нашёл кажется в исходниках, где это делается

_Файл exporter.js_
`function outputTracksWithAlert(source) {
    let tracks = [];
    for (i = 0; i < source.length; i++) {
        if (source[i].artists.length != 0 && source[i].title) {
            tracks.push('${source[i].artists[0].name} - ${source[i].title}');
`

Вот если ${source[i].artists[0].name} заменить на конкатенированную строку из всех имён исполнителей, то получится то, что надо.
Я бы и сам это сделал, но не знаю JavaScript...

Также подозреваю, что в source есть какой-то параметр помимо artists и title, в котором хранятся сведения о редакции трека.

**Comment by Chimildic** [comment]
Версия экспорта в расширении давно не развивалась. В сравнении с другими решениями, которые теперь не работают. 

> не сохраняется информация о редакции композиции (чей ремикс и т.д.)

Соглашусь, что потеря версии трека значимое упущение. По обратной связи никто не обращал на это внимание за все время существования расширения. 

> сохраняется информация только о первичном исполнителе

С исполнителями сложнее. По опыту из других проектов (например [goofy](https://github.com/Chimildic/goofy)), у Spotify очень кривой поиск. Далеко не гугл в своих решениях. Поэтому лишнее имя исполнителя или специальные знаки типа `&$#!` ломают поиск. Он даст пустой или некорректный результат. Хотя фактически может существовать трек даже с такими данными. 

Именно поэтому в указанному вами сайте на выходе был список треков, который частично очищался от специальных знаков. Для лучшего процента переноса нужно было даже еще больше сделать со строками. Поскольку за `spotlistr` и `tunemymusic` не замечано дополнительной помощи для улучшения поиска.

Хорошим решением бы стало реализация своего подключения к Spotify в расширении, чтобы контролировать весь процесс переноса. Но я не горю желанием заниматься этим в который раз. Вероятно предусмотрю дополнительные режимы формирования списка треков в расширении. Сроков пока не назову.

**Reply by KisoidPoh095** [reply]
Ок, ну в принципе, всё что я озвучил ранее для меня уже не так актуально, т.к. мне было нечем заняться и я написал свой скрипт на PowerShell, который всё экспортирует как мне надо :)
Вот скрипт, если кому интересно: https://pastebin.com/8Nybvheq
(кстати, непонятно, что же перестало работать у Яндекса, что старый метод экспорта по ссылке теперь недоступен? - я по идее тем же api пользовался...)
Так что смотрите сами...

Вашим расширением пока буду решать другие задачи. Уже проверил перемешивание треков в плейлисте и создание "дискографии" всех треков артиста - работает.

Вычитание треков из плейлиста у меня не заработало - сейчас сделаю отдельный реквест.

P.S.: Я использую экспорт для бэкапа плейлистов для надёжности, а не для Spotify (не доверяю на 100% облачным сервисам; особенно после закрытия некоторых, которыми активно пользовался раньше: Имхонета, Delicio.us, Wunderlist и т.д.)
Всю историю ластфм тоже бэкаплю с помощью https://benjaminbenben.com/lastfm-to-csv/

**Reply by AlexByte** [reply]
А вот странно, что spotlistr находит трек "bad things --- anybody" только если три тире добавить. У Spotify топ поиск ( особенно если вкладки Songs открывать, а не All), это у этих сайтов какая-то проблема с поиском. В Целом spotlistr получше будет, tunemymusic вообще левые треки подсовывает бывает, в 5% где-то, как при поиске avicii the days и OneRepublic Apologize и the killers mr brightside.
По результатам исследования из 500 треков spotlistr подобрал 8 левых треков, а tunemymusic подобрал 32 левых треков и ремиксов, то есть 1.6% против 6.4%. Но tunemymusic может добавлять неактивные треки в spotify.

**Comment by Chimildic** [comment, marked as answer]
Экспортер будет добавлять версию трека (не альбома) к названию с версии 0.8.0

**Comment by exentero** [comment]
А можно ли сделать два варианта экспорта плейлиста на выбор: для передачи в другие сервисы, как есть сейчас и второй с полным копированием номера трека, артиста, названия? Очень необходимая мне штука. Да и не только мне, думаю...

**Reply by kisoid** [reply]
Сомневаюсь, что сейчас кому-то вообще нужно экспортировать из ЯМузыки в Spotify (по крайней мере, живущим в РФ). А вот полный бэкап плейлиста (со всеми полями) - это задача актуальная в любом случае.

**Reply by exentero** [reply]
Почему именно Spotify? Мало ли кому куда нужно сделать экспорт... https://i.ibb.co/g65Tqbt/image.png

**Reply by kisoid** [reply]
Потому что экспорт в том виде, в котором он есть сейчас, всегда позиционировался как пригодный для импорта в Спотифай :) А про другие сервисы упоминаний не было никогда...
